node/tests/swamp/swamp.go

```diff
@@ -0,0 +1,191 @@
+package swamp
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"net"
+	"testing"
+	"time"
+
+	"github.com/ipfs/go-log/v2"
+	"github.com/libp2p/go-libp2p-core/host"
+	mocknet "github.com/libp2p/go-libp2p/p2p/net/mock"
+	ma "github.com/multiformats/go-multiaddr"
+	"github.com/stretchr/testify/require"
+
+	"github.com/celestiaorg/celestia-node/core"
+	"github.com/celestiaorg/celestia-node/libs/keystore"
+	"github.com/celestiaorg/celestia-node/node"
+	"github.com/celestiaorg/celestia-node/node/p2p"
+)
+
+var blackholeIP6 = net.ParseIP("100::")
+
+// Swamp represents the main functionality that is needed for the test-case:
+// - Network to link the nodes
+// - CoreClient to share between Bridge nodes
+// - Slices of created Bridge/Light Nodes
+// - trustedHash taken from the CoreClient and shared between nodes
+type Swamp struct {
+	t           *testing.T
+	Network     mocknet.Mocknet
+	CoreClient  core.Client
+	BridgeNodes []*node.Node
+	LightNodes  []*node.Node
+	trustedHash string
+}
+
+// NewSwamp creates a new instance of Swamp.
+func NewSwamp(t *testing.T) *Swamp {
+	if testing.Verbose() {
+		log.SetDebugLogging()
+	}
+
+	ctx := context.Background()
+	// TODO(@Bidon15): Rework this to be configurable by the swamp's test
+	// case, not here.
+	kvStore := core.CreateKvStore(200)
+	coreNode := core.StartMockNode(kvStore)
+
+	coreNode.Config().Consensus.CreateEmptyBlocksInterval = 200 * time.Millisecond
+
+	swp := &Swamp{
+		t:          t,
+		Network:    mocknet.New(ctx),
+		CoreClient: core.NewEmbeddedFromNode(coreNode),
+	}
+
+	swp.trustedHash = swp.getTrustedHash(ctx)
+
+	swp.t.Cleanup(func() {
+		swp.stopAllNodes(ctx, swp.BridgeNodes, swp.LightNodes)
+	})
+	return swp
+}
+
+// stopAllNodes goes through all received slices of Nodes and stops one-by-one
+// this eliminates a manual clean-up in the test-cases itself in the end
+func (s *Swamp) stopAllNodes(ctx context.Context, allNodes ...[]*node.Node) {
```

Comment:
optional nit: a bit weird to pass a 2d array while it can one only array.

Reply:
Well, there were 2 ideas in my head while opting for the one you are seeing here. 
1. If making this func without a 2d array, then there can be a repeated call of `stopAllNodes` in the cleanup, which is not what I wanted to see. 
2. This bit weird impl just simplifies adding future node types without the repeated calls

Reply:
Ok, let's keep this.

Reply:
I don't understand why this could not just be a 1d array either 🤔